cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/store/AppMetadataStoreTest.java

```diff
@@ -133,57 +136,193 @@ public void apply() throws Exception {
+    // KILLED status is persisted with the largest sourceId
+    assertPersistedStatus(metadataStoreDataset, txnl, 10L, 100L, 1000L, ProgramRunStatus.KILLED);
+    // STARTING status is persisted since no following sourceId is larger
+    assertPersistedStatus(metadataStoreDataset, txnl, 10L, 10L, 10L, ProgramRunStatus.STARTING);
+  }
+
+  @Test
+  public void testInvalidStatusPersistence() throws Exception {
+    final AppMetadataStore metadataStoreDataset = getMetadataStore("testInvalidStatusPersistence");
+    TransactionExecutor txnl = getTxExecutor(metadataStoreDataset);
+    // No status can be persisted if STARTING is not present
+    assertPersistedStatus(metadataStoreDataset, txnl, null,
+                          ImmutableList.of(ProgramRunStatus.RUNNING, ProgramRunStatus.COMPLETED));
+    assertPersistedStatus(metadataStoreDataset, txnl, null,
+                          ImmutableList.of(ProgramRunStatus.SUSPENDED, ProgramRunStatus.RESUMING));
+    assertPersistedStatus(metadataStoreDataset, txnl, null, STOP_STATUSES);
+    // Once a stop status is reached, any incoming status will be ignored
+    for (int i = 0; i < STOP_STATUSES.size(); i++) {
+      // A different stop status at the end will be ignored
+      assertPersistedStatus(metadataStoreDataset, txnl, STOP_STATUSES.get(i),
+                            ImmutableList.of(ProgramRunStatus.STARTING, STOP_STATUSES.get(i),
+                                             STOP_STATUSES.get((i + 1) % STOP_STATUSES.size())));
+    }
+    // STARTING status will be ignored if there's any existing record
+    assertPersistedStatus(metadataStoreDataset, txnl, ProgramRunStatus.SUSPENDED,
+                          ImmutableList.of(ProgramRunStatus.STARTING, ProgramRunStatus.SUSPENDED,
+                                           ProgramRunStatus.STARTING));
+    assertPersistedStatus(metadataStoreDataset, txnl, ProgramRunStatus.RUNNING,
+                          ImmutableList.of(ProgramRunStatus.STARTING, ProgramRunStatus.RUNNING,
+                                           ProgramRunStatus.STARTING));
+  }
+
+  private void assertPersistedStatus(final AppMetadataStore metadataStoreDataset, TransactionExecutor txnl,
+                                     final long startSourceId, final long runningSourceId,
+                                     final long killedSourceId, ProgramRunStatus expectedRunStatus) throws Exception {
     // Add some run records
-    TreeSet<Long> expected = new TreeSet<>();
     ApplicationId application = NamespaceId.DEFAULT.app("app");
     final ProgramId program = application.program(ProgramType.WORKFLOW, "program");
-    final RunId runId = RunIds.generate(10000);
-    // Program completed status is not recorded with a smaller sourceId
+    // use sourceId's to ensure runId is unique for each run
+    final RunId runId = RunIds.generate(10000 + startSourceId + runningSourceId + killedSourceId);
```

Comment:
Is there a need to add all these ids?

Reply:
Modified it to `RunIds.generate(startSourceId * 3 + runningSourceId * 5 + killedSourceId * 7)` to ensure that as long as startSourceId, runningSourceId and killedSourceId have a different combination, there is a unique runId